Add kdreg notifier to memory registration cache (and clean up)

sungeunchoi opened the pull request:
Add the option to use the kdreg IOMMU notifier to manage memory
registrations that have been unmapped by the OS.  Every memory region
is monitored upon successful registration (fi_mr_reg) and unmonitored
upon deregistration (fi_close).  The notifier is checked for events
that identify monitored memory regions that have had 1 or more pages
unmapped.  Unmapped regions are marked as such and removed from use
(retired) or deleted from the cache.  Retired memory registrations are
entries that have not been closed (i.e., deregistered) and may be due
to a user freeing memory before closing the memory registration.  A
warning is emitted in this case.

Some minor restructuring and clean up was also done with this commit.

- Added notifier functionality to the cache, to be used with or
  without kdreg.  TODO: further generalize to plug in any notifier.

- Added __mr_cache_lru_remove() and replaced all dlist function on the
  lru list with the appropriate __mr_cache_lru_* functions (for
  clarity's sake).

- Other minor stuff:
  - Removed legacy check for entry match in __mr_cache_flush
  - Improved error checking
  - Replaced asserts with calls to GNIX_FATAL
  - Added/updated GNIX_INFO output
  - Updated comments and doxygen annotations

- gnitest changes:
  - Added new notifier tests
  - Changed rdm_sr, api, and api_cq tests to close memory regions
    before freeing them
  - Cleaned up mr tests a bit (got rid of unused stuff and made stuff
    file static)
  - Added valgrind suppression for reading kernel managed counter

KNOWN BUGS:

- Stale entry that becomes unmapped during __mr_cache_search_inuse,
  can result in double free
- Entry in to_destroy list can be freed by notifier call

@jswaro @hppritcha 

Signed-off-by: Sung-Eun Choi <sungeunchoi@users.noreply.github.com>

jswaro commented on prov/gni/src/gnix_mr.c:
```diff
@@ -212,20 +212,9 @@ DIRECT_FN int gnix_mr_reg(struct fid *fid, const void *buf, size_t len,
-		}
-	}
-
-	rc = _gnix_mr_cache_register(domain->mr_cache,
```
Why did this get removed? I believe that Zach added this to address a performance issue, but the details are a little hazy.

sungeunchoi replied:
It got moved into _gni_mr_cache_register (and restructured a little).

jswaro replied:
This seems a bit dubious. A user shouldn't call into a structure if it hasn't been initialized and now we've migrated the initialization point into the operation. Additionally, this might interfere with close behavior, but I'm not sure at this point. 

I'd advise against taking external locks in a provider agnostic implementation such as we are trying to migrate to.

sungeunchoi replied:
It's only GNI provider code that would call into the wrapper I'm talking, so I wouldn't exactly call it user code.  The MR interface would go back to what it used to be.

jswaro commented on prov/gni/src/gnix_mr_cache.c:
```diff
@@ -81,20 +85,19 @@ typedef struct gnix_mr_cache_entry {
 /* forward declarations */
-int _gnix_mr_cache_init(
-		gnix_mr_cache_t      **cache,
-		gnix_mr_cache_attr_t *attr);
```
This seems to be going in the opposite direction of being a general purpose memory registration cache if we intend to use this with other providers later.

sungeunchoi replied:
Agreed.  I can put an inline wrapper in there.  I preferred the simpler call site to this function.

sungeunchoi replied:
FYI, removed this forward declaration, since it is not needed (already in the .h).

jswaro commented on prov/gni/src/gnix_mr_cache.c:
```diff
@@ -130,7 +136,7 @@ gnix_mr_cache_attr_t __default_mr_cache_attr = {
  * @param[in] y key to be compared against
  *
- * @return    -1 if it should be positioned at the left, 0 if the same,
+ * @return    -1 if it should be positioned at the left, 0 if fully enclosed,
```
0 if the entry overlaps

sungeunchoi replied:
Got it.

jswaro commented on prov/gni/src/gnix_mr_cache.c:
```diff
@@ -259,7 +265,11 @@ static inline void __attach_retired_entries_to_registration(
-	assert(dlist_empty(retired_entries));
+
+	if (!dlist_empty(retired_entries)) {
+		GNIX_FATAL(FI_LOG_MR, "retire_entries not empty\n");
```
retired_entries

sungeunchoi replied:
Check.

jswaro commented on prov/gni/src/gnix_mr_cache.c:
```diff
@@ -337,6 +349,236 @@ static inline int __mr_cache_lru_dequeue(
+				 * problem.*/
+				if (__notifier_warned == false) {
+					GNIX_WARN(FI_LOG_MR,
+						  "In use memory registration"
```
In-use?

sungeunchoi replied:
Actually, since it's a warning to the end user, it shouldn't use internal implementation terminology.  Maybe something more direct like "Registered memory region includes unmapped regions".

jswaro commented on prov/gni/src/gnix_mr_cache.c:
```diff
@@ -337,6 +349,236 @@ static inline int __mr_cache_lru_dequeue(
+						       iter);
+					if (ret != RBT_STATUS_OK) {
+						GNIX_FATAL(FI_LOG_MR,
+							   "Umapped entry"
```
unmapped

sungeunchoi replied:
Got it.

jswaro commented on prov/gni/src/gnix_mr_cache.c:
```diff
@@ -337,6 +349,236 @@ static inline int __mr_cache_lru_dequeue(
+				ret = rbtErase(cache->stale.rb_tree, iter);
+				if (ret != RBT_STATUS_OK) {
+					GNIX_FATAL(FI_LOG_MR,
+						   "Umapped entry could"
```
unmapped

sungeunchoi replied:
Check.

jswaro commented on prov/gni/src/gnix_mr_cache.c:
```diff
@@ -373,6 +622,15 @@ static inline int __insert_entry_into_stale(
+	if ((entry->state == GNIX_CES_INUSE_UNMAPPED) ||
+	    (entry->state == GNIX_CES_STALE_UNMAPPED)) {
+		GNIX_INFO(FI_LOG_MR, "entry (%p) unmapped, not inserting"
+			  " into stale %llx:%llx", entry,
```
Could we simply make UNMAPPED a flag rather than a state? It doesn't feel like a state. This is a nit. It doesn't really affect how we do things here at the moment.

sungeunchoi replied:
I have a note up top that says we should reevaluate how we maintain states and flags, possibly combining all in a single state variable.  Let's talk more about it later.

jswaro commented on prov/gni/src/gnix_mr_cache.c:
```diff
@@ -441,25 +713,42 @@ static inline void __resolve_stale_entry_collision(
-		assert(!ret);
+		if (ret) {
+			GNIX_FATAL(FI_LOG_MR,
+				   "Failed to insert submsumed MR "
```
subsumed

sungeunchoi replied:
Check.

jswaro commented on prov/gni/src/gnix_mr_cache.c:
```diff
@@ -767,12 +1070,20 @@ int __mr_cache_flush(gnix_mr_cache_t *cache, int flush_count) {
 		}
 
+		/* sungeun: I'm confused as to why this works in the
+		 * case where e_entry != entry */
```
Legacy code I believe.

sungeunchoi replied:
Thx, will remove comment.

sungeunchoi replied:
And the legacy code too.

jswaro commented on prov/gni/src/gnix_mr_cache.c:
```diff
@@ -943,6 +1268,10 @@ static int __mr_cache_search_stale(
 	gnix_mr_cache_entry_t *mr_entry, *tmp;
 
+	if (cache->attr.lazy_deregistration) {
+		__clear_notifier_events(cache);
```
Is the notifier only ever used in lazy deregistration? I'd assume this would affects all versions of the cache, since a region could be unmapped and the user could attempt to register a subsection of an existing registration that has been partially unmapped.

sungeunchoi replied:
I think you're right here.  I have a mental note to reevaluate how often we check for events, but maybe it would be good to check even without lazy deregistration  sometimes.  I'll create an issue so we can discuss.

sungeunchoi commented:
@jswaro Addressed comment from original review.  __gnix_mr_cache_[de]register is back to the way it was (didn't seem worth wrapping for a single call).

TODOs based the discussions:
- Review use of flags and states
- Review when __clear_notifier_events is called

jswaro commented on prov/gni/include/gnix_mr_cache.h:
```diff
@@ -220,10 +221,18 @@ int _gnix_mr_cache_destroy(gnix_mr_cache_t *cache);
  */
 int _gnix_mr_cache_init(
 		gnix_mr_cache_t      **cache,
+		struct gnix_mr_notifier *mr_notifier,
```
I'd like this to be a memory of the attributes that way if a notifier isn't specified, we can handle it in the init, rather than changing the API.

sungeunchoi replied:
Good idea.

jswaro replied:
There is a nice side benefit to putting this in the attributes and checking to see if it has been provided or not. If it hasn't been provided, we can either use the default notifier (stubbed and returned -FI_EBUSY) or just check to see if it exists when necessary (if (cache->notifier) do_something(event)). This should keep it pretty agnostic as well.

--Jim

---

From: Sung-Eun Choi notifications@github.com
Sent: Monday, May 9, 2016 6:19:55 PM
To: ofi-cray/libfabric-cray
Cc: James Swaro; Mention
Subject: Re: [ofi-cray/libfabric-cray] Add kdreg notifier to memory registration cache (and clean up) (#809)

In prov/gni/include/gnix_mr_cache.hhttps://github.com/ofi-cray/libfabric-cray/pull/809#discussion_r62589686:

>   _/
>  int _gnix_mr_cache_init(
>               gnix_mr_cache_t      *_cache,
> -             struct gnix_mr_notifier *mr_notifier,

Good idea.

## 

You are receiving this because you were mentioned.
Reply to this email directly or view it on GitHubhttps://github.com/ofi-cray/libfabric-cray/pull/809/files/a98a4782dc49e4ddd84f1b37eb6f373b51800594#r62589686

sungeunchoi commented:
@jswaro Notifier moved to the attributes.

jswaro commented on prov/gni/src/gnix_mr_cache.c:
```diff
@@ -508,11 +796,18 @@ static inline int __mr_cache_entry_put(
+
+	if (cache->attr.lazy_deregistration) {
+		__clear_notifier_events(cache);
+	}
```
Is this going to handled in a later PR? I can't quite recall. 

This is related to the discussion that notifier events may pertain to registrations in a non-lazy cache as well.

jswaro commented on prov/gni/src/gnix_mr_notifier.c:
```diff
@@ -0,0 +1,245 @@
+	}
+}
+
+#endif /* HAVE_KDREG */
```
I know this seems trivial, but it might make sense to have tests run for without KDREG to ensure that the default values are still returned for the dummy monitor (as we expect them to).

sungeunchoi replied:
I agree.  I plan to have it run without kdreg for a bit before flipping the switch.  Then I'll have a small subset of tests run both with and without.

jswaro commented on prov/gni/test/api.c:
```diff
@@ -247,6 +242,11 @@ static void rdm_api_teardown_common(bool unreg)
+	free(uc_target);
+	free(target);
+	free(source);
+
```
Why did this move?

sungeunchoi replied:
Because we're freeing the memory before closing the MR.  (Commit message includes this).

jswaro commented on prov/gni/test/api_cq.c:
```diff
@@ -247,6 +242,11 @@ static void api_cq_teardown_common(bool unreg)
+	free(uc_target);
+	free(target);
+	free(source);
+
```
Same question as above. Why did this move?

hppritcha commented on prov/gni/configure.m4:
```diff
@@ -102,6 +102,14 @@ dnl looks like we need to get rid of some white space
+		    gni_CPPFLAGS="-I$with_kdreg/include -DHAVE_KDREG $gni_CPPFLAGS"
+        fi
+
+
```
we should replace this with pkg-config m4 code at some point.

sungeunchoi replied:
We talked about this before, right?  This is not possible until the package is actually available on a system.  Or is there something else I'm missing?

hppritcha commented:
the configury changes need major help, but I can do that in a subsequent PR.
:+1:
